OutgoingWebSocket.js

```diff
@@ -0,0 +1,199 @@
+      }
+    });
+    this.registerListeners();
+  }
+
+  /**
+   * Registers the socket listeners.
+   */
+  registerListeners() {
+    this.socket.on('open', this.openHandler.bind(this));
+    this.socket.on('message', this.messageHandler.bind(this));
+    this.socket.on('close', this.closeHandler.bind(this));
+    this.socket.on('error', this.errorHandler.bind(this));
+  }
+
+  /**
+   * Triggers when socket connection is opened.
+   */
+  openHandler() {
+    if (this.reconnectInfo !== null) {
+      logger.debug(`${OUTGOING} [${this.connectionId}] [RECONNECT] - ${this.reconnectInfo}`);
+      this.send(this.reconnectInfo);
+    }
+    this.emit(kConnectionOpened);
+    this.shouldRetry = false;
+    this.retryCount = config.retryVal;
+  }
+
+  /**
+   * Triggers when message is received on socket.
+   *
+   * @param {string} msg
+   */
+  messageHandler(msg) {
+    if (msg != null && msg != undefined && msg != "" && msg.substring(0, 9) === RECONNECT) {
+      this.reconnectInfo = msg;
+      this.emit(kAddNewContext, this.connectionId);
+      this.emit(kReleaseTap);
+      return;
+    }
+    this.emit(kMessageReceived, msg);
+  }
+
+  /**
+   * Triggers when socket connection is closed.
+   *
+   * @param {number} code
+   * @param {string} msg
+   */
+  closeHandler(code, msg) {
+    if (!this.shouldRetry) {
+      if (msg === SERVICE_RESTART) {
+        this.shouldRetry = true;
+        this.emit(kUpstreamRestart, code, msg);
+        this.startRetries(code, msg);
+      }
+      else {
+        this.emit(kUpstreamClosed, code, msg);
+      }
+    }
+  }
+
+  /**
+   * Triggers when error occured on socket.
+   */
+  errorHandler() {
+    this.emit(kError);
+  }
+
+  /**
+   * Retries upstream socket until max retries reached.
+   *
+   * @param {number} code
+   * @param {string} msg
+   */
+  async startRetries(code, msg) {
+    if (this.shouldRetry) {
+      if (this.retryCount == 0) {
+        this.emit(kUpstreamClosed);
+      } else {
+        this.retryCount = this.retryCount - 1;
+        await sleep(config.retryDelayVal);
+        this.addSocket();
+      }
+      logger.debug(`${OUTGOING} [${this.connectionId}] [RETRIES LEFT: ${this.retryCount}] `);
+    }
+  }
+
+  /**
+   *  Closes the socket connection.
+   */
+  close() {
+    this.socket.close();
+  }
+
+  /**
+   * Adds message to queue.
+   *
+   * @param {string} msg
+   */
+  addToQueue(msg) {
+    this.queue.enqueue(msg);
+  }
+
+  /**
+   * Sends the message on socket.
+   *
+   * @param {string} msg
+   */
+  send(msg) {
+    this.socket.send(msg);
+  }
+
+  /**
+   * Drains the queue and emits completed event.
+   */
+  drainQueue() {
+    while (!this.queue.isEmpty()) {
+      this.emit(kDrainMessage, this.queue.dequeue());
+    }
+    this.emit(kDrainCompleted);
+  }
+
+  /**
+   * Sets connection identifier from headers
+   */
+  setConnectionId() {
+    this.connectionId = extractConnectionId(this.headers);
+  }
```

Comment:
Is this getting used?

Reply:
yes we set connection id in outgoing socket & used for logging pupose